cmd/soroban-rpc/internal/db/transaction.go

```diff
@@ -0,0 +1,387 @@
+	NewTx(ctx context.Context) (TransactionReaderTx, error)
+	GetLedgerRange(ctx context.Context) ledgerbucketwindow.LedgerRange
+}
+
+// TransactionReaderTx provides all of the public ways to read from the DB.
+// Note that `Done()` *MUST* be called to clean things up.
+type TransactionReaderTx interface {
+	GetTransaction(log *log.Entry, hash xdr.Hash) (Transaction, bool, ledgerbucketwindow.LedgerRange)
+	GetLedgerRange() ledgerbucketwindow.LedgerRange
+	Done() error
+}
+
+type transactionHandler struct {
+	db         db.SessionInterface
+	stmtCache  *sq.StmtCache
+	passphrase string
+
+	ingestMetric, insertMetric, countMetric prometheus.Observer
+}
+
+type transactionReaderTx struct {
+	ctx        context.Context
+	db         db.SessionInterface
+	passphrase string
+}
+
+func NewTransactionReader(db db.SessionInterface, passphrase string) TransactionReader {
+	return &transactionHandler{db: db, passphrase: passphrase}
+}
+
+func NewTransactionWriter(db db.SessionInterface, stmtCache *sq.StmtCache, passphrase string) TransactionWriter {
+	return &transactionHandler{db: db, stmtCache: stmtCache, passphrase: passphrase}
+}
+
+func (txn *transactionHandler) InsertTransactions(lcm xdr.LedgerCloseMeta) error {
+	start := time.Now()
+	txCount := lcm.CountTransactions()
+	L := log.
+		WithField("ledger_seq", lcm.LedgerSequence()).
+		WithField("tx_count", txCount)
+
+	if txn.stmtCache == nil {
+		return errors.New("TransactionWriter incorrectly initialized without stmtCache")
+	}
+
+	if txCount == 0 {
+		L.Warnf("No transactions present in ledger: %+v", lcm)
+		return nil
+	}
+
+	reader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(txn.passphrase, lcm)
+	if err != nil {
+		return errors.Wrapf(err,
+			"failed to open transaction reader for ledger %d",
+			lcm.LedgerSequence())
+	}
+
+	transactions := make(map[xdr.Hash]ingest.LedgerTransaction, txCount)
+	for i := 0; i < txCount; i++ {
+		tx, err := reader.Read()
+		if err != nil {
+			return errors.Wrapf(err, "failed reading tx %d", i)
+		}
+
+		// For fee-bump transactions, we store lookup entries for both the outer
+		// and inner hashes.
+		if tx.Envelope.IsFeeBump() {
+			transactions[tx.Result.InnerHash()] = tx
+		}
+		transactions[tx.Result.TransactionHash] = tx
+	}
+
+	mid := time.Now()
+	L.WithField("passphrase", txn.passphrase).
+		Debugf("Ingesting %d transaction lookups from ledger", len(transactions))
+
+	query := sq.Insert(transactionTableName).
+		Columns("hash", "ledger_sequence", "application_order", "is_soroban")
+	for hash, tx := range transactions {
+		hexHash := hex.EncodeToString(hash[:])
+		is_soroban := (tx.UnsafeMeta.V == 3 && tx.UnsafeMeta.V3.SorobanMeta != nil)
+		query = query.Values(hexHash, lcm.LedgerSequence(), tx.Index, is_soroban)
+	}
+	_, err = query.RunWith(txn.stmtCache).Exec()
+
+	L.WithError(err).
+		WithField("duration", time.Since(start)).
+		WithField("sql_duration", time.Since(mid)).
+		Infof("Ingested %d transaction lookups", len(transactions))
```

Comment:
I don't think this is worth logging because we have metrics reporting this info

Reply:
True but metrics are aggregated - there have been many instances where it's been very useful to see individual timelines for certain cases in Horizon